packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts

```diff
@@ -786,20 +786,78 @@ const transferDateSectionValue = <TDate>(
   }
 };
 
+const getCleanISOSectionValue = (value: number) => {
+  return value < 10 ? `0${value}` : value.toString();
+};
+
+const getCleanISOYearValue = (value: number) => {
+  let result;
+  if (value < 10) {
+    result = `000${value}`;
+  } else if (value < 100) {
+    result = `00${value}`;
+  } else if (value < 1000) {
+    result = `0${value}`;
+  }
+  return result ? result.toString() : value.toString();
+};
+
+const transferDateSectionsValue = <TDate>(
+  utils: MuiPickersAdapter<TDate>,
+  sections: FieldSectionWithoutPosition[],
+  dateToTransferFrom: TDate,
+  dateToTransferTo: TDate,
+) => {
+  const yearSection = sections.find((section) => section.type === 'year');
+  const monthSection = sections.find((section) => section.type === 'month');
+  const daySection = sections.find((section) => section.type === 'day');
+  const fromJSDate = utils.toJsDate(dateToTransferFrom);
+  const toJSDate = utils.toJsDate(dateToTransferTo);
+  const parsedDate = utils.date(
+    new Date(
+      `${getCleanISOYearValue(
+        yearSection ? fromJSDate.getFullYear() : toJSDate.getFullYear(),
+      )}-${getCleanISOSectionValue(
+        (monthSection ? fromJSDate.getMonth() : toJSDate.getMonth()) + 1,
+      )}-${getCleanISOSectionValue(daySection ? fromJSDate.getDate() : toJSDate.getDate())}`,
```

Comment:
With this strategy, I'm a bit scarred about timezone.

> When parsing date strings with the Date constructor (and Date.parse, they are equivalent), always make sure that the input conforms to the ISO 8601 format (YYYY-MM-DDTHH:mm:ss.sssZ) — the parsing behavior with other formats is implementation-defined and may not work across all browsers.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date


Another solution would be to fix section order. For example always doing `['year', 'month', 'day']` would allows to prevent the bug and stick with getter/setter of the date library

Reply:
Yeah, for me the whole point of the current strategy is to not loose information like time, timezone, locale etc... of the date.

Couldn't we just merge into the date in a predetermined order (year => month => ...) instead of doing it in the section order ?

Reply:
> Yeah, for me the whole point of the current strategy is to not loose information like time, timezone, locale etc... of the date.

I can think of another pro for going with something like the suggested approach—less steps (date modifications).


> Couldn't we just merge into the date in a predetermined order (year => month => ...) instead of doing it in the section order ?

I have thought of it as well, but haven't experimented with it. Will look into this option. 😉

Reply:
My assumption was that, if we only push info inside the given date using `setYear`, `setHours` etc... then we lower the risk of loosing info we don't even know exist and that can be library-specific.
Any creation of date from scratch is in a sense a risk.

But of course, that's a trade-off that must be challenge if the downside are too big.